keps/sig-storage/1710-selinux-relabeling/README.md

```diff
@@ -51,223 +48,175 @@
 
-For CSI, kubelet uses following heuristics:
+For CSI, kubelet currently uses following heuristics:
 
-1. Mount the volume (via `NodeStage` + `NodePublish` CSI calls).
-2. Check mount options of the volume mount dir. If and only if it contains `seclabel` mount option, the volume supports SELinux.
+1. Mount the volume as usual (via `NodeStage` + `NodePublish` CSI calls).
+2. Check mount options of the volume mount dir. If and only if it contains `seclabel` mount option, the volume supports SELinux and kubelet will pass ":Z" to the container runtime, which then relabels all files there.
 
 ### Goals
 
-Optionally (chosen by user), do not recursively relabel content of the volumes.
+* If possible, mount volumes with the correct SELinux context using `-o context=XYZ` mount option and avoid recursive change of all files on the volume.
 
 ### Non-Goals
 
-Change container runtimes / CRI.
+* Change container runtimes / CRI.
 
 ## Proposal
 
-Offer option in `Pod.Spec.PodSecurityContext to` *mount* volumes with the right labels instead of recursive relabeling:
-
-```go
-type SELinuxRelabelPolicy string
-
-const (
-    OnVolumeMount SELinuxRelabelPolicy = "OnVolumeMount"
-    AlwaysRelabel SELinuxRelabelPolicy = "Always"
-)
-
-type PodSecurityContext struct {
-    // seLinuxRelabelPolicy ← new field
-    // Defines behavior of changing SELinux labels of the volume before being exposed inside Pod.
-    // Valid values are "OnVolumeMount" and "Always". If not specified, "Always" is used.
-    // "Always" policy recursively changes SELinux labels on all files on all volumes used by the Pod.
-    // "OnVolumeMount" tries to mount volumes used by the Pod with the right context and skip recursive ownership
-    // change. Kubernetes may fall back to policy "Always" if a storage backed does not support this policy.
-    // This field is ignored for Pod's volumes that do not support SELinux.
-    // + optional
-    SELinuxRelabelPolicy *SELinuxRelabelPolicy
+* When kubelet *knows* SELinux context of a pod (i.e. at least `Pod.Spec.SecurityContext.SELinuxOptions.Level` is set) AND kubelet *knows* that a volume supports mounting with `-o context`:
+  * kubelet passes `-o context=XYZ` to `MountDevice()` and `SetUp()` calls of the volume.
+    * A volume plugin / CSI driver will get these as regular mount options and use them to mount the volume.
+  * kubelet does not pass any special SELinux option to the container runtime (explicitly, no `:Z`).
+    Files on the volume already have the right SELinux context, no recursive relabeling happens there.
+  * When a `PodSecurityContext` specifies incomplete SELinux label (i.e. omits `SELinuxOptions.User`, `.Role` or `.Type`), kubelet fills the blanks from the system defaults provided by [ContainerLabels() from go-selinux bindings](https://github.com/opencontainers/selinux/blob/621ca21a5218df44259bf5d7a6ee70d720a661d5/go-selinux/selinux_linux.go#L770).
```

Comment:
Above we say we are going to shorten the selinux label when we apply it via `mount -o context` and here we are saying user/role/type is obtained via go-selinux bindings?

Reply:
The shortening `system_u:system_r:container_t:s0:c309,c383` -> `so:c309,c383` was meant only to make the text more readable, all `mount -o context` would be always with the full context, incl. `system_u:system_r:container_t`. Since it did not save much text, I removed the shortening and all examples have the full context / label.